### PR TITLE
MAINT: Translate selection.c.src to C++ using templates.

### PR DESCRIPTION
**Pull request by serge-sans-paille**
This one was a bit more difficult to convert due to the extra template loop on arg/ noarg.

### CODE REVIEW

File: numpy/core/src/npysort/selection.c.src [hunk 1]
```diff
@@ -1,419 +0,0 @@
...
- *         LONGLONG, ULONGLONG, HALF, FLOAT, DOUBLE, LONGDOUBLE,
- *         CFLOAT, CDOUBLE, CLONGDOUBLE#
- * #suff = bool, byte, ubyte, short, ushort, int, uint, long, ulong,
- *         longlong, ulonglong, half, float, double, longdouble,
- *         cfloat, cdouble, clongdouble#
- * #type = npy_bool, npy_byte, npy_ubyte, npy_short, npy_ushort, npy_int,
- *         npy_uint, npy_long, npy_ulong, npy_longlong, npy_ulonglong,
- *         npy_ushort, npy_float, npy_double, npy_longdouble, npy_cfloat,
- *         npy_cdouble, npy_clongdouble#
- * #inexact = 0*11, 1*7#
```
Comment (serge-sans-paille) [inline numpy/core/src/npysort/selection.c.src:72]: this inexact looks odd to me: if I understand the syntax correctly, it means the 11 first entries are exact (0) but then `npy_ushort` is flagged as inexact (1) for no apparent reason? I updated that during the conversion.

**Comment by serge-sans-paille** [comment]
(based on #20829 for no good reason)

File: numpy/core/src/npysort/selection.c.src (hunk 1 above)
Comment (charris) [reply numpy/core/src/npysort/selection.c.src:72]: The 16 bit floats are stored in 16 bit unsigned integers, the compiler doesn't (didn't?) have a half type.

Comment (charris) [reply numpy/core/src/npysort/selection.c.src:72]: As a result, arithmetic for half floats is done in  float32, with conversion before and after. Comparisons may also be tricky.

**Comment by charris** [comment]
It is taking me a while to wrap my head around this, with what looks like a mix of template declarations and definitions. The specialization of `Sortee<typename type, true>` looks almost like inheritance, is that how it works? I'm a bit curious how well this performs with the operator overloading. I assume you used `()` instead of `[]` because you wanted to use the last in it's normal form?

**Comment by charris** [comment]
Thanks @serge-sans-paille .

**Pull request merged**